packages/discovery-react-components/src/components/DocumentPreview/components/PdfViewer/PdfViewer.tsx

```diff
@@ -88,33 +118,74 @@ const PdfViewer: SFC<Props> = ({
     };
   }, [loadedFile, page]);
 
-  const [viewport, canvasInfo] = useMemo(() => {
-    const viewport = loadedPage?.getViewport({ scale });
-    const canvasInfo = viewport ? getCanvasInfo(viewport) : undefined;
-    return [viewport, canvasInfo];
-  }, [loadedPage, scale]);
+  const currentPage = useMemo(() => {
+    const isPageValid = !!loadedPage && loadedPage.pageNumber === page;
+    if (isPageValid) {
+      const viewport = loadedPage?.getViewport({ scale });
+      const canvasInfo = viewport ? getCanvasInfo(viewport) : undefined;
+      return { loadedPage, viewport, canvasInfo };
+    }
+    return null;
+  }, [loadedPage, page, scale]);
 
   useEffect(() => {
-    if (loadedPage && !loadedPage.then && viewport && canvasInfo) {
-      _renderPage(loadedPage, canvasRef.current!, viewport, canvasInfo);
-      setLoading(false);
+    let didCancel = false;
+    let task: PDFRenderTask | null = null;
+
+    const { loadedPage, viewport, canvasInfo } = currentPage || {};
+    if (loadedPage && !(loadedPage as any).then && viewport && canvasInfo) {
+      const render = async () => {
+        try {
+          task = _renderPage(loadedPage, canvasRef.current!, viewport, canvasInfo);
+          await task?.promise;
+        } catch (e) {
+          if (e instanceof RenderingCancelledException) {
+            // ignore
+          } else {
+            throw e; // rethrow unknown exception
+          }
+        } finally {
+          if (!didCancel) {
+            setLoading(false);
+          }
+        }
+      };
+      render();
     }
-  }, [loadedPage, viewport, canvasInfo, setLoading]);
+    return () => {
+      didCancel = true;
+      task?.cancel();
+    };
+  }, [loadedPage, currentPage, setLoading]);
 
   useEffect(() => {
     if (setHideToolbarControls) {
       setHideToolbarControls(false);
     }
   }, [setHideToolbarControls]);
 
+  const classNameBase = `${settings.prefix}--document-preview-pdf-viewer`;
+  const { loadedPage: currentLoadedPage, canvasInfo } = currentPage || {};
```

Comment:
Again, this seems unnecessary to rename `loadedPage` here -- just use the existing var.